### PR TITLE
docs: update lease section

### PR DESCRIPTION
**Pull request by tbg**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9645)

### CODE REVIEW

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/design.md, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwPXCiKRGX9Im9sYpR:-KSwPXCiKRGX9Im9sYpS:bkvx1rw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L134)):*
> ```Markdown
> Each range maintains a small (i.e. latest 10s of read timestamps),
> *in-memory* cache from key to the latest timestamp at which the
> key was read. This *read timestamp cache* is updated every time a key
> ```

not a "read" timestamp cache. it's also not the only case when this thing gets updated (delrange)

---

*[docs/design.md, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwPfk-Sy4G31JsoFc4:-KSwPfk0GK1EccAW6p3d:bb7qx3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L804)):*
> ```Markdown
> (i.e. stopping traffic completely for inactive ranges).
> 
> # Replica Leases
> ```

wat?

---

*[docs/design.md, line 839 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwPlSebVkxSdz_vhiw:-KSwPlSebVkxSdz_vhix:bjt1507) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L839)):*
> ```Markdown
> the former lease holder). This is accomplished by letting leases enter
> a <i>stasis period</i> (which is just the expiration minus the maximum clock
> ofsset) before the actual expiration of the lease, so that all the next lease
> ```

ofssssssseetttt

---

*[docs/design.md, line 854 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwQPLPVHVmlkFlMQnI:-KSwQPLPVHVmlkFlMQnJ:bltaon5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L854)):*
> ```Markdown
> 
> The range lease is completely separate from Raft leadership, and so without
> further efforts, Raft leadership and the Range lease may not be represented by the same. Since it's expensive to not have these two roles colocated, each lease
> ```

the same..what?

---

*[docs/design.md, line 856 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwQQm2-8bIWAcFgX8A:-KSwQQm2-8bIWAcFgX8B:bp97202) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L856)):*
> ```Markdown
> further efforts, Raft leadership and the Range lease may not be represented by the same. Since it's expensive to not have these two roles colocated, each lease
> renewal or transfer also attempts to colocate them. In practice, that means
> that this phenomenon is rare and self-corrects quickly.
> ```

you should probably make more of a point of defining this "phenomenon" if you're going to refer to it so casually. this section is titled pretty generally, not "raft leadership and range lease non-locality".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645)*

File: docs/design.md [hunk 1]
```diff
@@ -834,28 +834,26 @@ overhead of going through Raft.
...
-Raft leader, which finally commits it into the log and updates its followers,
-including the Range lease holder. This yields correct results but wastes several
-round-trip delays, and so we will make sure that in the vast majority of cases
-Range lease and Raft leadership coincide. A fairly easy method for achieving this is
-to have each new lease period (extension or new) be accompanied by a
-stipulation to the lease holder's replica to start Raft elections (unless it's
-already leading), though some care should be taken that Range lease holdership is
-relatively stable and long-lived to avoid a large number of Raft leadership
-transitions.
+further efforts, Raft leadership and the Range lease may not be represented by the same. Since it's expensive to not have these two roles colocated, each lease
```
Comment (petermattis) [inline docs/design.md]: Can you elaborate on the expense? Something along the lines of `the lease holder needs to forward write operations to the leader`.

File: docs/design.md [hunk 2]
```diff
@@ -803,12 +801,14 @@ batch processing of requests.
...
 (i.e. stopping traffic completely for inactive ranges).
 
-# Range Leases
+# Replica Leases
 
 As outlined in the Raft section, the replicas of a Range are organized as a
 Raft group and execute commands from their shared commit log. Going through
 Raft is an expensive operation though, and there are tasks which should only be
 carried out by a single replica at a time (as opposed to all of them).
+In particular, it is desirable to serve authoritative reads from at least one
```
Comment (petermattis) [inline docs/design.md]: `at least one` sounds odd. `from a single Replica`?

File: docs/design.md [hunk 3]
```diff
@@ -803,12 +801,14 @@ batch processing of requests.
 Future optimizations may include two-phase elections and quiescent ranges
 (i.e. stopping traffic completely for inactive ranges).
 
-# Range Leases
+# Replica Leases
```
Comment (petermattis) [inline docs/design.md]: I thought `Range Leases` was the correct term.

File: docs/design.md [hunk 4]
```diff
@@ -834,28 +834,26 @@ overhead of going through Raft.
 Since reads bypass Raft, a new lease holder will, among other things, ascertain
 that its timestamp cache does not report timestamps smaller than the previous
 lease holder's (so that it's compatible with reads which may have occurred on
-the former lease holder). This is accomplished by setting the low water mark of the
-timestamp cache to the expiration of the previous lease plus the maximum clock
-offset.
+the former lease holder). This is accomplished by letting leases enter
+a <i>stasis period</i> (which is just the expiration minus the maximum clock
+ofsset) before the actual expiration of the lease, so that all the next lease
```
Comment (petermattis) [inline docs/design.md]: `s/ofsset/offset/g`

File: docs/design.md (hunk 1 above)
Comment (petermattis) [inline docs/design.md]: `s/by the same/by the same Replica/g`

**Review by petermattis — APPROVED** [review]

**Comment by tschottdorf** [comment]
PTAL

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions.

---

*[docs/design.md, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwPXCiKRGX9Im9sYpR:-KSwuubtO6zrqcFPaNvq:b-yi1opf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L134)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not a "read" timestamp cache. it's also not the only case when this thing gets updated (delrange)
</details>
I agree, but that's outside of the section. I don't see the point in talking about the timestamp cache in the "Versioned Values" section in the first place.

---

*[docs/design.md, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwO26IxH70PZLJQ1NE-r1-804:-KSwwbbLduMecxNZrYIg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L804)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I thought `Range Leases` was the correct term.
</details>
Done.

---

*[docs/design.md, line 804 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwPfk-Sy4G31JsoFc4:-KSwwbvj2Xj3O6X5nBPu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L804)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wat?
</details>
Done.

---

*[docs/design.md, line 810 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwO26IxH70PZLJQ1NE-r1-810:-KSwwkOHkxHXwzSheppl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L810)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `at least one` sounds odd. `from a single Replica`?
</details>
Done.

---

*[docs/design.md, line 839 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwO26IxH70PZLJQ1NE-r1-839:-KSwwlzPsACCBCnru8qN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L839)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/ofsset/offset/g`
</details>
Done.

---

*[docs/design.md, line 839 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwPlSebVkxSdz_vhiw:-KSwwmWWELPIBBuMhws0:b72386x) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L839)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ofssssssseetttt
</details>
http://vignette2.wikia.nocookie.net/disney/images/8/81/Junglebook-disneyscreencaps_com-6045.jpg/revision/latest?cb=20150917164058

---

*[docs/design.md, line 854 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwO26IxH70PZLJQ1NE-r1-854:-KSwxBMX0cDeMgkr6cPI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L854)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/by the same/by the same Replica/g`
</details>
Done.

---

*[docs/design.md, line 854 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwQPLPVHVmlkFlMQnI:-KSwxBfsUqpWNyM9wcop:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L854)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the same..what?
</details>
Done.

---

*[docs/design.md, line 856 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwQQm2-8bIWAcFgX8A:-KSwxJJVSFGmQ1y0bDv_:b-gk5vq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L856)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you should probably make more of a point of defining this "phenomenon" if you're going to refer to it so casually. this section is titled pretty generally, not "raft leadership and range lease non-locality".
</details>
Done, I think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-:-KSwxmxrCH28oSLjudA9:bb71pro)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-:-KSwyIAgBorDz-A0DwCe:bnfp4nl)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[docs/design.md, line 134 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwPXCiKRGX9Im9sYpR:-KSwyMiK5-JPFKI3yjMN:bnd6xtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L134)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I agree, but that's outside of the section. I don't see the point in talking about the timestamp cache in the "Versioned Values" section in the first place.
</details>
:gun: 

---

*[docs/design.md, line 854 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwO26IxH70PZLJQ1NE-r1-854:-KSwz6TLEY04HzSUNu_t:b6hapwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L854)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Done.
</details>
"may not be held" is ambiguous. "might not be held"?

---

*[docs/design.md, line 186 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwyQ8YWr63KypuB80y:-KSwyQ8YWr63KypuB80z:b-rjmako) ([raw file](https://github.com/cockroachdb/cockroach/blob/583fe4c6b8fde0c0bbd2d6cc09a0d154ab240f36/docs/design.md#L186)):*
> ```Markdown
> Both SI and SSI require that the outcome of reads must be preserved, i.e.
> a write of a key at a lower timestamp than a previous read must not succeed. To
> this end, each range maintains a small (i.e. latest 10s of read timestamps),
> ```

just say it's bounded

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645)*

**Comment by tschottdorf** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[docs/design.md, line 854 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwO26IxH70PZLJQ1NE-r1-854:-KSxaOu5ZIAApDcJwuL6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9ca16748a72db02881ca0b136fcc7a438b4e86e/docs/design.md#L854)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> "may not be held" is ambiguous. "might not be held"?
</details>
Done.

---

*[docs/design.md, line 186 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645#-KSwyQ8YWr63KypuB80y:-KSxaIGg1XCofCAl17bl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/583fe4c6b8fde0c0bbd2d6cc09a0d154ab240f36/docs/design.md#L186)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> just say it's bounded
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9645)*

**Pull request merged**